src/lib/integration_test_lib/util.ml

```diff
@@ -49,6 +50,11 @@ let run_cmd_exn dir prog args =
   | Error error ->
       Error.raise error
 
+let run_cmd_exn_timeout ~timeout_seconds dir prog args =
+  Timeout.await ()
+    ~timeout_duration:(Time.Span.create ~sec:timeout_seconds ())
+    (run_cmd_exn dir prog args)
+
```

Comment:
@nholland94  i'm pretty sure there's something else i need to make this function work but i don't know what it is

Reply:
This will work in order to make the program stop waiting forever for the child process to finish. However, this implementation isn't ideal, because it won't kill the child process when the timeout occurs, so if we have a loop that keeps running new commands (which we have in a few places), this will end up spawning a bunch of child processes. To kill the child process when the timeout occurs, you would need to write some custom logic here to manage the process differently than how `run_cmd` manages it right now, since in the current model, you just return a deferred that resolves when the command completes. What you would want to do is keep a reference to the process returned by `Process.create_exn` so that you can send it a kill signal if the timeout occurs.

Reply:
If you separate the logic to handle the output of a process into another function `check_cmd_output : Process.Output.t -> string Or_error.t` so you can share it across this function and the existing `run_cmd_or_error` function, something like the following pseudocode might work.

```ocaml
[%log' spam (Logger.create ())]
    "Running command (from %s): $command" dir
    ~metadata:[ ("command", `String (String.concat (prog :: args) ~sep:" ")) ] ;
let process = Process.create_exn ~working_dir:dir ~prog ~args () in
match%map
  Timeout.await ()
    ~timeout_duration:(Time.Span.create ~sec:timeout_seconds ())
    (Process.collect_output_and_wait process)
with
| `Ok output -> check_cmd_output output
| `Timeout -> Or_error.errors "timed out running command"
```